### ISSUE TITLE
get rid of default value specification in command-line parsing API?

### ISSUE THREAD

**Issue by jdtournier** [MRtrix3/mrtrix3#334]
At the moment, the `.type_integer()` & `.type_float()` methods take a default value, but this value is actually never used anywhere in the code - in fact, it's not even accessible in the API... 

It was initially designed to be used to build a GUI for command-line only commands - the idea was to create an app that would run its target command with the `__print_full_usage__` argument, which causes the app to dump essentially all the information in the `usage()` function. The GUI app would then build a GUI from that information, and use the defaults provided to initialise fields where possible. The user would then fill out the information, press run, and the GUI app would invoke the command-line app. 

Only problem is that I never got round to building the GUI part - and I'm not sure I'll ever do it. So unless someone else thinks this would be worthwhile, and wants to invest the time into this, maybe we should just remove those default values? Otherwise it's just confusing for new developers to be providing information that isn't used anywhere...

**Comment by draffelt** [comment]
Rami is currently using it for bash completion after choosing an option (ie
you hit tab after an option and it gives you the default). However given
the reason you would normally provide an option is to not use the default
value, I think it's safe to go.

**Comment by jdtournier** [comment]
I guess the issue isn't whether it's safe to go - I'm pretty sure it is. It's more whether we have any plans to use these values at any point in the future - with the only application I can foresee being this GUI idea...

**Comment by Lestropie** [comment]
I must admit I've thought about the GUI command interface on multiple occasions. But it's difficult to justify the time. And given how trivial it would be to re-introduce default values if that were to be pursued, it's probably better to reduce the confusion in current usage.

**Comment by thijsdhollander** [comment]
I'm also all for (currently) removing the default values.  I must admit I've myself been confused by them quite some time, wondering why they were there.  I also think the GUI thing in this case actually doesn't add *that* much to the usability; as long as the commands (try to) do a good job at producing meaningful (error) messages when things are (going) wrong, or provided values/parameters are wrong, combined with a good help text for each option, we should be good.

**Comment by jdtournier** [comment]
OK, guess we have a consensus. I also agree the GUI thing is pretty redundant, but then I wouldn't be the one using it... Anyway, I'll get this done once the updated_syntax branch is merged.

**Comment by dchristiaens** [comment]
Sorry I'm a bit late to the party, but I would like to offer a counter-argument. I too have been annoyed at the non-use of the default arguments for a while. For example, my GT code is full of things like:
```
  + Option ("t0",
            "set the initial temperature of the metropolis hastings optimizer.")
    + Argument ("start").type_float(1e-6, 0.1, 1e6)

// ...

  double t0 = 0.1;
  opt = get_options("t0");
  if (opt.size())
    t0 = opt[0][0];
```
Instead of getting rid of unused defaults, why not actually _use_ them? Then, the example above could be reduced to:
```
double t0 = get_options("t0")[0][0];
```
or even nicer (adding a function to the API):
```
double t0 = get_option_value("t0");
```
I'm aware that this won't work for all options (e.g., there's no such thing as a _default_ response function, although...), but at least for those numeric parameters it could make life easier. It would avoid mistakingly specifying different defaults in the code and docs (or having to define macros for all defaults), and it would allow to set defaults for recurring options such as `lmax` MRtrix-wide instead of in each command separately.

P.S.: Note that in the code above, the user still cannot look up the default without digging in the source code. This means that if I want to document default numeric values, I need to specify them a _third_ time as part of the text description of the option. If kept, it might be a good idea to include them in the docs as well.

**Comment by bjeurissen** [comment]
I agree with Daan. It would be nice to expose the default values and actually use them further down in the command implementation.

**Comment by jdtournier** [comment]
@dchristiaens, that was more or less the plan from day one: this API is not a great deal different from the 0.2.x version, and default values were in there too at that time (back in ~2006). There's a few reasons why they never really went beyond that. 

The lazy reason is that the overwhelming majority of the time, the code responsible for managing the option is very close to the code using the option (i.e. in the same app, or in the same cpp file). It's really not that much effort to write a macro to state the default, and use the macro in those few places (or just to put the actual value itself) - it's certainly less effort than updating the API to allow these defaults, making sure all the other issues this would introduce are handled appropriately. But the main point really is that I've never managed to conjure up a nice, clean and simple way to implement this that didn't introduce other issues of one form or another... but then maybe I just haven't given that much thought... Besides, the number of options this relates to is actually relatively small - only integer, float, and list types can have defaults that make any sense. 

An _actual_ reason is that very often, there is no such thing as a default, even for an int or float type - the default value instead depends on other parameters or input data (as I mentioned earlier today in #132). Or there genuinely is no default - setting the option changes the algorithm in such as way that it now does require that value, but that value would be unused otherwise. So you'd need an API flexible enough to _allow_ default parameters, but not require them. 

Yet another reason relates to the documentation. As I mentioned in #132, the description for arguments to options is ignored, for reasons explained in that post. The better place to document these arguments is within the description for the parent option. While it would be possible for the help page generation code to state the defaults immediately after the option's description, I think this would be sufficiently clumsy enough of the time as to make it undesirable. For instance, you might have an option taking a number of arguments, each with their own defaults. It might make sense to have these arguments described in the option description, in a way that isn't a straight bullet point list - i.e. the description of what the option does clarifies what each argument is about, but you wouldn't necessarily go: argument1 does this, argument2 does that, etc. Listing the default values automatically afterwards may feel pretty cluncky, or duplicate information that's actually already stated in the option description. In the previous version of MRtrix, I often struggled with this kind of issue, how to describe argument X knowing that it would be better explained in the context of its option rather than as a standalone bit of text afterwards. It meant I often resorted to slightly awkward descriptions so as to avoid duplicating a lot of the information. Now, this relates to the _description_ of the arguments, but I'm pretty sure it would apply to the default values too. To put it another way: having the defaults (or description) listed automatically by the backend would force the developer to describe thing differently from the way they'd actually prefer. And it's not that big a deal to just document these things manually.

I should add, in relation to the MRtrix-wide options such as "lmax": there is kind of a mechanism to do this via OptionGroups - at least any MRtrix-wide functions that are aware of these options can be made aware of the default. This is the case particularly for the "lmax" option, which is parsed by a number of different functions all within the same .h/.cpp pair as the OptionGroup itself. Where I do agree with you though is that this doesn't mean other applications that might want to parse those options themselves would be aware of the defaults... But then, for lmax this is already more complicated than a single constant, since it would depend on the input data... 

So in short, while I agree this would be nice in principle, I reckon that in practice, the defaults would end up getting in the way often enough to get irritating - and worse, introduce user-visible changes to the way commands are documented that are actually not what the developer would consider optimal. I think going forward I'd rather have any relevant defaults specified as a `constexpr` or macro, and placed in the relevant header where that makes sense. 

As to the reduction in lines of codes, I reckon this could be achieved by adding a simple function to the existing API along the lines that @dchristiaens was suggesting, e.g.:

    double t0 = get_option_value("t0", t0_default_value);

with the only difference being that you'd actually provide the default in that function, rather than having the backend figure it out from the option specification. This should be sufficiently convenient, but also allow that default value to be computed prior to that line for those cases where it might depend on other inputs. Bear in mind that this particular syntax is only appropriate in a subset of cases: when there is only one argument to the option; there is no additional code that needs to be run if that option is set; and only a single instance of that option is allowed. For just about any other case, you'd want to resort to the current API anyway.

All this to say, the more I think about it, the more I think this default entry needs to go...

**Comment by jdtournier** [comment]
In relation to @dchristiaens's suggestion of having syntax like:

    double t0 = get_options("t0")[0][0];

This would not fit with the current API: the `get_options()` call returns a vector of the options that were actually specified on the command-line. If you were to use the default value, this implies no such option was provided, and that vector would therefore be empty, so the subsequent subscript operators would crash out. 

@dchristiaens's second option is however feasible - for those cases where it is appropriate.

**Comment by dchristiaens** [comment]
> Besides, the number of options this relates to is actually relatively small - only integer, float, and list types can have defaults that make any sense.

Well, a quick count by Qt Creator gave me over 160 usages of integer, float, and sequence types. Sure, that's a bit smaller than the no. image and file types (190), but given those don't have defaults it's not really a fair comparison.

I do see your point though. Some of these defaults are specified in terms of _something_ else, be it the voxel size, data quality, or some other parameter. It would be difficult to define their defaults in the command usage or option group, let alone document their values automatically... Regarding the OptionGroup mechanism, they are certainly very handy but I have always felt they violate encapsulation a bit, e.g., the CSDeconv class should not be parsing command line options (and if it wouldn't, you could build a GUI that didn't have to invoke the command-line app). Besides, "lmax" is used in at least 3 other commands out of the scope of the CSD OptionGroup (obviously...), so it still doesn't set one single default. Anyhow, that's a different discussion...

But ultimately, we're after the same thing: getting these defaults set up and documented with as little code duplication as possible. If that means killing the default values in the options/arguments, like you suggested, and think of those purely as a means of IO and documentation, I can live with that :smile: A `get_option_value` function to set the default would be nice though. I would be happy with something along the lines of your suggestion:
```
+ Option ("step", "set the step size of the algorithm in mm (default is 0.1 x voxelsize).")
          + Argument ("size").type_float (0.0, INFINITY)

// ...

float stepsize = get_option_value("step", 0.1 * voxsize);
```
Clean, concise, and readable :wink: If we have a consensus, I'll add such function to the API.

**Comment by jdtournier** [comment]
Glad to hear you can see my point. More than happy for you to add that `get_option_value()` function to the API, I think it'll come in handy for a lot of people. Best to add it to the `updated_syntax` branch only at this point...

I can see you point about the OptionGroup and encapsulation. Ultimately though, it's not so much an issue with the OptionGroup per se, but with the way they're used in the backend. It would be possible to split the command-line parsing aspects from the CSDeconv class itself if needed, simply by adding a function (maybe a static member of the class?) to do the parsing and returning a fully-formed CSDeconv instance, for example. And this may very well be a better way to do things. Definitely happy for this to change if anyone thinks it's a problem. 

In any case, I assume this means everyone is happy with the removal of the defaults? If so, I'll get rid of them in the `updated_syntax` branch at some point when I have the time...

**Comment by bjeurissen** [comment]
I am happy with it.

**Comment by thijsdhollander** [comment]
Yep, I also still support the removal of the defaults; the `get_option_value` function is indeed welcome though, but given its quite "specific" requirements as mentioned before, its usage will still be more exception than rule I expect (but that's perfectly fine).

Once every command/option than *can* benefit of it is updated to do so, you could actually do a more accurate count to get a view on how prevalent such options are. :wink:

**Comment by dchristiaens** [comment]
I've added that function and refactored the most obvious cases in commits 3c0c769 and 3a573fe. For those who like accurate counts, that's 42 usages already, only 2 of which were initialised by an expression instead of a default value. This eliminates 112 lines of code, and that's only for those options set within the command. I did not yet dig into the src-folder.

**Comment by jdtournier** [comment]
:+1:

**Comment by jdtournier** [comment]
I think we still need to remove the default value from the `.type_*()` calls...?

**Comment by Lestropie** [comment]
See `args_no_default_values` branch for removal of default values.

Ironically, a lot of the work for this went into updating default value prompts in the help text, which is precisely what the default value variables could have been used for...

@dchristiaens: Only conflict I found was in the cpot and ppot variables in `tckglobal`: the default values quoted in the help text didn't match the actual values used. Have a look and make sure you're happy with the values before this gets merged.

**Comment by jdtournier** [comment]
Well that must have been fun...

**Comment by dchristiaens** [comment]
@Lestropie Good catch Rob, fixed it in 933d01a. That proves that you've done some very useful work making these defaults explicit...

**Comment by jdtournier** [comment]
Closing, see pull request #533.
